### PR TITLE
VM-scoped persistent ResourceClaims for DRA device passthrough

### LINKED ISSUE #18171: Support VM-scoped persistent ResourceClaims for DRA device passthrough
## Summary

ResourceClaims created for DRA device passthrough (GPUs, NVMe drives, NICs) are currently pod-scoped — they are deleted when the virt-launcher pod dies. Restarting a VM means a new claim and potentially a different physical device. This is problematic for:

- **NVMe drives** — data loss when a different device is allocated
- **GPUs** — driver incompatibility, licensing issues on device change
- **Busy clusters** — race condition where another workload grabs the device during restart

## Proposal

Add a `resourceClaimTemplates` field to `VirtualMachineSpec` (following the established `dataVolumeTemplates` pattern) so that ResourceClaims are owned by the VM instead of the VMI pod. The VM controller creates ResourceClaims from templates with the VM as controller owner, ensuring device allocations persist across VM restarts.

Key changes:
- **API**: New `ResourceClaimTemplateEntry` type and `ResourceClaimTemplates` field on `VirtualMachineSpec`
- **Controller**: `handleResourceClaims()` creates claims during `startVMI()`, informer event handlers for reconciliation
- **VMI setup**: `SetupVMIFromVM` rewrites `resourceClaimTemplateName` references to direct `resourceClaimName` references before VMI creation
- **Webhook**: Validates feature gate, duplicate names, required fields, matching VMI resourceClaims entries

## Prior Art

- [GSoC 2024: Persistent Device Claims for KubeVirt](https://github.com/kubevirt/community/issues/254)
- [Design Doc](https://docs.google.com/document/d/1wThUzwNhfhdM5YxtqxwrATsFx2eMTeircFEme1Fzrd0/edit)

## Implementation

PR: https://github.com/kubevirt/kubevirt/pull/17957

/kind feature
/sig compute

### PR DESCRIPTION
**Pull request by johnahull**
Fixes https://github.com/kubevirt/kubevirt/issues/18171

## Summary

- Adds `resourceClaimTemplates` field to `VirtualMachineSpec`, enabling ResourceClaims to be owned by the VM instead of the VMI pod
- VM controller creates ResourceClaims from ResourceClaimTemplates with the VM as controller owner, ensuring device allocations persist across VM restarts
- `SetupVMIFromVM` rewrites `resourceClaimTemplateName` references to direct `resourceClaimName` references before VMI creation
- Follows the established `dataVolumeTemplates` pattern for consistency

## Motivation

When using DRA with KubeVirt today, ResourceClaims are pod-scoped — they're deleted when the virt-launcher pod dies. Restarting a VM means a new claim and potentially a different physical device. This is problematic for:

- **NVMe drives** — data loss when a different device is allocated
- **GPUs** — driver incompatibility, licensing issues on device change
- **Busy clusters** — race condition where another workload grabs the device during restart

This PR makes device assignments persist across VM restarts by scoping ResourceClaims to the VM lifecycle, the same way `dataVolumeTemplates` scopes DataVolumes.

## Prior Art

- [GSoC 2024: Persistent Device Claims for KubeVirt](https://github.com/kubevirt/community/issues/254)

## Changes

- **API**: New `ResourceClaimTemplateEntry` type and `ResourceClaimTemplates` field on `VirtualMachineSpec`
- **Controller**: `handleResourceClaims()` creates claims during `startVMI()`, informer event handlers for reconciliation, expectation tracking
- **VMI setup**: Rewrites template references to direct claim names
- **Webhook**: Validates feature gate, duplicate names, required fields, matching VMI resourceClaims entries
- **RBAC**: virt-controller permissions for resourceclaims and resourceclaimtemplates
- **Example**: `examples/vm-dra-pgpu.yaml`
- **Tests**: Unit tests for claim rewriting, manifest building, and edge cases

## Design Doc

[VEP #344: VM-Scoped Persistent ResourceClaims](https://github.com/kubevirt/enhancements/pull/345) ([enhancement issue](https://github.com/kubevirt/enhancements/issues/344))

## Test plan

- [x] Unit tests for `SetupVMIFromVM` ResourceClaim rewriting (single, multiple, mixed, edge cases)
- [x] Unit tests for `CreateResourceClaimManifest` ownership and naming
- [x] Unit tests for webhook validation
- [ ] Integration tests for full VM lifecycle with persistent claims
- [ ] E2E tests with a real DRA driver

```release-note
Added `resourceClaimTemplates` to VirtualMachineSpec, enabling DRA ResourceClaims to persist across VM restarts.
```

### CODE REVIEW

**Comment by johnahull** [comment]
All three sourcery-ai suggestions addressed in c44a475:

1. Extracted `ResourceClaimNameForVM()` helper in `util.go`, used by all three call sites.
2. Refactored `validateResourceClaimTemplates`: precomputed template claim names, added error when `spec.Template` is nil.
3. Added unit tests for `handleResourceClaims`: existing claim, creation, missing template, failed creation.

**Comment by johnahull** [comment]
Friendly ping. This is ready for review when someone has bandwidth.

**Comment by jean-edouard** [comment]
> Friendly ping. This is ready for review when someone has bandwidth.

/hold
The VEP needs to be approved and merged first, I'll have a look

File: pkg/virt-operator/resource/generate/rbac/controller.go [hunk 1]
```diff
@@ -589,8 +589,37 @@ func baseControllerClusterRole() *rbacv1.ClusterRole {
...
 					"resourceclaims",
 				},
+				Verbs: []string{
+					"list",
+					"watch",
+					"get",
+					"create",
+					"update",
+					"patch",
+					"delete",
```
Comment (jean-edouard) [inline pkg/virt-operator/resource/generate/rbac/controller.go]: That's a lot of permissions! I don't see them all used, could you please detail why each of them is needed?

Comment (johnahull) [reply pkg/virt-operator/resource/generate/rbac/controller.go]: Good catch, I over-provisioned. Trimmed to what's actually used:

- resourceclaims: `list`, `watch`, `get` (informer for the claim store) + `create` (`handleResourceClaims` creates claims from templates). Removed `update`, `patch`, `delete`: owner references handle GC, and the controller never modifies claims after creation.
- resourceclaimtemplates: `list`, `watch`, `get` (informer for the template store). Controller reads templates to create claims.
- resourceslices: removed entirely — not used by virt-controller.

Will push the fix.

File: examples/vm-dra-pgpu.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
+---
+# This example demonstrates VM-scoped persistent ResourceClaims using DRA.
+# Unlike vmi-dra-pgpu.yaml (which uses pod-scoped claims that are deleted on restart),
+# this VM creates ResourceClaims owned by the VM. The claims persist across VMI restarts,
+# ensuring the same device is re-attached when the VM is restarted.
+#
+# Prerequisites:
+#   - Kubernetes DRA (DynamicResourceAllocation) feature gate enabled
+#   - KubeVirt GPUsWithDRA or HostDevicesWithDRA feature gate enabled
```
Comment (jean-edouard) [inline examples/vm-dra-pgpu.yaml]: And PersistentDRAClaimsGate?

File: pkg/virt-controller/watch/application.go [hunk 3]
```diff
@@ -419,6 +420,8 @@ func Execute() {
 	app.allPodInformer = app.informerFactory.Pod()
 	app.exportServiceInformer = app.informerFactory.ExportService()
 	app.resourceQuotaInformer = app.informerFactory.ResourceQuota()
+	app.resourceClaimInformer = app.informerFactory.ResourceClaim()
+	app.resourceClaimTemplInformer = app.informerFactory.ResourceClaimTemplate()
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/application.go]: A lot of code is missing the feature gate guard. Only the webhook checks `PersistentDRAClaims`, but these informers (and the `hasSynced`, expectations, the event handlers they feed and more) run unconditionally.

**Review by jean-edouard — COMMENTED** [review]
Probably needs a `make generate && make`.
More code needs gating, see in-line comment.
/test all

File: examples/vm-dra-pgpu.yaml (hunk 2 above)
Comment (johnahull) [reply examples/vm-dra-pgpu.yaml]: Added in b488c40 — the example now lists `PersistentDRAClaims` as a prerequisite.

File: pkg/virt-controller/watch/application.go (hunk 3 above)
Comment (johnahull) [reply pkg/virt-controller/watch/application.go]: Fixed in b488c40. ResourceClaim and ResourceClaimTemplate informers are now gated behind `PersistentDRAClaims` using the same dummy-informer pattern as CDI/DataVolumes:

```go
if app.isPersistentDRAClaimsEnabled {
    app.resourceClaimInformer = app.informerFactory.ResourceClaim()
    app.resourceClaimTemplInformer = app.informerFactory.ResourceClaimTemplate()
} else {
    app.resourceClaimInformer = app.informerFactory.DummyResourceClaim()
    app.resourceClaimTemplInformer = app.informerFactory.DummyResourceClaimTemplate()
}
```

When the gate is off, dummy informers are passed to `NewController`: empty stores, no watches, `HasSynced` returns immediately, expectations are never populated. The controller wires up normally but never sees any ResourceClaim objects. Reinit triggers if the gate is toggled at runtime.

**Comment by johnahull** [comment]
Updated with significant changes based on review feedback and hardware testing:

**New: `persistWhenStopped` field** on `ResourceClaimTemplateEntry`
- `false` (default): claims deleted on explicit stop, freeing devices for other workloads
- `true`: claims persist with allocation intact across stop/start — same physical device guaranteed
- Claims always persist across reboots regardless of this setting

**New: `reservedFor` allocation preservation**
When `persistWhenStopped: true`, the controller adds the VM to the claim's `status.reservedFor`. This prevents the DRA resource controller from deallocating the device when the pod is deleted on stop. Without this, the claim goes to `pending` and may get a different device on restart.

**New: `PersistentDRAClaims` feature gate** (alpha, disabled by default)
- Dedicated gate instead of reusing `GPUsWithDRA`/`HostDevicesWithDRA`
- Webhook rejects `resourceClaimTemplates` when gate is off

**New: name length validation**
- Derived claim name (`<vm-name>-<entry-name>`) validated against 63-char limit

**RBAC changes:**
- Added `resourceclaims/binding: patch` (required to modify `reservedFor`)
- Removed unused `update`/`patch`/`delete` → now: `get, list, watch, create, delete`
- Removed `resourceslices` entirely

**Hardware validation (Dell XE9680, 4x Samsung PM1745 NVMe via DRA VFIO):**

| Test | VMs | Cycles | Result |
|---|---|---|---|
| `persistWhenStopped=true` stop/start | 4 | 10 | 10/10 — same device preserved |
| `persistWhenStopped=true` reboot | 4 | 10 | 10/10 — same device preserved |
| `persistWhenStopped=false` stop/start | 4 | 10 | 10/10 — claims deleted/recreated |
| `persistWhenStopped=false` reboot | 4 | 10 | 10/10 — claims persist across reboot |

VEP updated in PR #345 to document `reservedFor` mechanism and RBAC.

File: pkg/virt-controller/watch/application.go (hunk 3 above)
Comment (jean-edouard) [reply pkg/virt-controller/watch/application.go]: Not fixed

File: manifests/generated/operator-csv.yaml.in [hunk 4]
```diff
@@ -848,12 +848,27 @@ spec:
...
           verbs:
           - list
           - watch
           - get
+          - create
+          - delete
+        - apiGroups:
+          - resource.k8s.io
+          resources:
+          - resourceclaims/binding
```
Comment (jean-edouard) [inline manifests/generated/operator-csv.yaml.in:870]: Did you mean
```suggestion
          - resourceclaims/status
```
?

File: pkg/virt-controller/watch/vm/vm.go [hunk 5]
```diff
@@ -572,6 +599,188 @@ func (c *Controller) handleDataVolumes(vm *virtv1.VirtualMachine) (bool, error)
...
+			continue
+		}
+		log.Log.Object(vm).Infof("Adding VM to ResourceClaim %s reservedFor (persistWhenStopped=true)", claimName)
+		if err := c.addVMToClaimReservedFor(claim, vm); err != nil {
+			log.Log.Object(vm).Warningf("Failed to add VM to ResourceClaim %s reservedFor: %v", claimName, err)
+		}
+	}
+}
+
+func (c *Controller) removeVMFromAllClaimReservedFor(vm *virtv1.VirtualMachine) {
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vm/vm.go:779]: Missing unit test coverage of the `reservedFor` machanism?

**Review by jean-edouard — COMMENTED** [review]
Please rework the commits to help reviewers. Things like "chore: regenerate API artifacts after rebase onto main" and "fix: address review feedback for VM-scoped persistent ResourceClaims" (and others) don't make sense on their own and should be part of the commits that introduce the things they're fixing.

File: pkg/virt-controller/watch/application.go (hunk 3 above)
Comment (johnahull) [reply pkg/virt-controller/watch/application.go]: Fixed in `1ea25506ec`. The feature is now gated in three layers:

1. **Webhook** — `ValidateVirtualMachineSpec` rejects `resourceClaimTemplates` when `PersistentDRAClaims` is off, so the field cannot be set in the first place.
2. **Informers** — the `ResourceClaim` / `ResourceClaimTemplate` informers are dummy informers when the gate is off (`application.go`), with a runtime reinit if the gate is toggled. No real watches, `HasSynced`, expectations, or event handlers fire.
3. **Reconcile methods** — `handleResourceClaims`, `ensureVMInClaimReservedFor`, `removeVMFromAllClaimReservedFor`, and `cleanupNonPersistentResourceClaims` each early-return when the gate is disabled.

The specific gap you flagged here was the VMI ref rewrite: it used to live inside `SetupVMIFromVM`, keyed only on field presence, so it ran regardless of the gate. It is now extracted into `rewriteResourceClaimTemplateRefs(vmi, vm)` and gated at its sole caller (`startVMI`, behind `PersistentDRAClaimsEnabled()`). `SetupVMIFromVM` is a pure VM to VMI projection again, so its 30 other callers are unaffected.

(The branch was rebased and squashed to 2 logical commits, which is why this anchor points at an older line.)

File: manifests/generated/operator-csv.yaml.in (hunk 4 above)
Comment (johnahull) [reply manifests/generated/operator-csv.yaml.in:870]: Not a typo. We intentionally grant both subresources:

- `resourceclaims/status` is the real write path (your suggestion, kept).
- `resourceclaims/binding` is a synthetic subresource the apiserver additionally SAR-checks whenever a status write changes `status.reservedFor` or `status.allocation`. See `AuthorizedForBinding` in k8s 1.36 (`pkg/registry/resource/utils.go`), gated by the `DRAResourceClaimGranularStatusAuthorization` feature gate. Because the controller adds a non-Pod (`virtualmachines`) consumer to `reservedFor`, this second check applies and would deny the write without the grant.

One change since your review: the verb is now `update` (was `patch`). `reservedFor` is a shared list also written by the scheduler, so the controller writes it via a live `Get` + `UpdateStatus` under `RetryOnConflict` rather than a merge patch (a merge patch would replace the whole list and clobber concurrently added consumers). `UpdateStatus` is an HTTP PUT, authorized as `update`, and the binding SAR uses the same verb as the request, so both rules are `update`. Regenerated in `1ea25506ec`.

File: pkg/virt-controller/watch/vm/vm.go (hunk 5 above)
Comment (johnahull) [reply pkg/virt-controller/watch/vm/vm.go:779]: Added in `1ea25506ec`. There is now a `reservedFor management` context in `vm_test.go` covering:

- `ensureVMInClaimReservedFor` status-updates the VM in when the claim is allocated
- no-op when `persistWhenStopped` is false
- no-op when the claim is not allocated yet
- no-op when the VM is already in `reservedFor`
- skips a same-named claim owned by another VM
- no-op when the claim is not cached
- `removeVMFromAllClaimReservedFor` status-updates the VM out
- no-op when the VM is absent from `reservedFor`

Each case asserts the write goes through `UpdateStatus` on the `status` subresource (the get reactor returns the live claim, the update reactor captures the pushed object) so the `reservedFor` contents are verified, not just the call. The pure `reservedFor` helpers (`hasVMInReservedFor`) also have direct table tests in `resourceclaim_test.go`.

(The branch was rebased and squashed to 2 logical commits, which is why this anchor points at an older line.)

**Comment by johnahull** [comment]
Thanks for the review, @jean-edouard. I rebased on `main` and squashed the branch to 2 logical commits (`c91768e07f` API + `1ea25506ec` virt-controller impl), so the older review anchors have shifted. I have replied inline to each of the three open threads (feature-gate guarding, the `resourceclaims/status` vs `resourceclaims/binding` RBAC, and the `reservedFor` unit-test coverage). One note carried over into the replies: the `reservedFor` write now uses `UpdateStatus` + `RetryOnConflict` instead of a merge patch, so the RBAC verb on `resourceclaims/status` and `resourceclaims/binding` changed from `patch` to `update` (regenerated). Note `/hold` is still on intentionally until the VEP merges.
